tests/integration_test.rs

```diff
@@ -19,54 +19,106 @@ use predicates::prelude::*; // Used for writing assertions
 use std::process::Command; // Run programs
                            // use assert_cmd::cmd::Command; // Run programs - it seems to be equal to "use assert_cmd::prelude::* + use std::process::Command"
 
+use once_cell::sync::Lazy;
+use regex::bytes::Regex;
+use rusoto_core::Region;
+use rusoto_dynamodb::{DynamoDb, DynamoDbClient};
 use std::fs::File;
 use std::io::{self, Write}; // Used when check results by printing to stdout
+use std::sync::Mutex;
+use std::time::Duration;
+use tokio::time::delay_for;
 
 use tempfile::Builder;
 
 /// Integration tests would go with DynamoDB Local, so before running them setup() starts up DynamoDB Local with Docker.
 /// FYI: https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBLocal.html
 ///      https://hub.docker.com/r/amazon/dynamodb-local
-fn setup() -> Result</* std::process::Command */ Command, Box<dyn std::error::Error>> {
-    setup_with_port(8000)
+async fn setup() -> Result</* std::process::Command */ Command, Box<dyn std::error::Error>> {
+    setup_with_port(8000).await
 }
 
-fn setup_with_port(port: i32) -> Result<Command, Box<dyn std::error::Error>> {
-    // NOTE: setup() spins up DynamoDB Local in 8000 port as "local" region assumes it. Better to make it configurable.
-    let mut docker = Command::new("docker");
+// We use std::sync::Mutex instead of tokio::sync::Mutex, because mutex must be poisoned after setup failure.
+static SETUP_MUTEX: Lazy<Mutex<i32>> = Lazy::new(|| Mutex::new(0));
 
-    let check_cmd = docker.args(&[
+/// Check existence of docker process for dynamodb-local
+async fn check_dynamodb_local_running(port: u16) -> bool {
+    let mut docker_for_check = Command::new("docker");
+
+    let check_cmd = docker_for_check.args(&[
         "ps",
-        "-q",
-        "--filter",
-        &format!("expose={}", port),
+        "--format",
+        "{{.Ports}}",
         "--filter",
         "ancestor=amazon/dynamodb-local",
     ]);
-    let check_out = check_cmd.output().expect("failed to execut check cmd");
-    if check_out.stdout.len() != 0 {
-        println!("DynamoDB Local is already running.");
+    let check_out = check_cmd.output().expect("failed to execute check cmd");
+    let reg_str = format!(r"(?m):{}->\d+/tcp$", port);
+    let port_re = Regex::new(&reg_str).unwrap();
+    if !check_out.status.success() {
+        panic!("failed to execute docker ps command")
+    }
+    if port_re.is_match(&check_out.stdout) {
+        true
+    } else {
+        false
+    }
+}
+
+async fn setup_with_port(port: i32) -> Result<Command, Box<dyn std::error::Error>> {
+    // Check the current process at first to allow multiple threads to run tests concurrently
+    if check_dynamodb_local_running(port as u16).await {
         return Ok(Command::cargo_bin("dy")?);
     };
```

Comment:
We don't need to call `check_dynamodb_local_running()` twice, I think.

Assume that the first `check_dynamodb_local_running()` still exists here.
All multi-threaded tests run `check_dynamodb_local_running()` once for each at first.
When the docker container has been already created, they can return soon. That's good!
But when the docker container hasn't been created yet, the first test locks the mutex and creates a new docker container.
All the others need to wait it and then run `check_dynamodb_local_running()` again.

Then, assume that the first `check_dynamodb_local_running()` is removed.
Multi-threaded tests try to lock the mutex first, the only first one can lock and the others need to wait.
At first glance, this seems to lead to slow down.
But, when the docker container has already been created, the first one runs `check_dynamodb_local_running()` once and the mutex is **unlocked soon**, and then the others restart and run `check_dynamodb_local_running()` only once for each.
When the docker container hasn't been created yet, the first one runs `check_dymanodb_local_running()` once and creates a new docker container, and then the others run`check_dymanodb_local_running()` once for each.

Please let me know if you have any opinions and discussions.

Reply:
As I describe in https://github.com/awslabs/dynein/pull/28#issuecomment-972880324, this is a intended implementation for performance reasons. Although, I will add a comment for later reference.
Please let me know if you have any other question regarding this implementation.